Editor: Multi-Entity Saving UI Improvements

ockham opened the pull request:
## Description
Closes part of #31456.

- Update panel header
- Add entity descriptions to site and template sections
- Sort entities (site before templates before template types before content)

## How has this been tested?
- Use the TT1 Blocks theme.
- Go to the Site Editor, and select the Single Post Template to edit.
- Make a few minor changes: Change the site title and tagline; insert a paragraph block with some text near the post content; change the address in the site footer.
- Click 'Save'.

## Screenshots

| Before | After |
|-------|-------|
|![image](https://user-images.githubusercontent.com/96308/138758681-40f3c871-6796-4b51-96e9-f7a60deb45bb.png)|![image](https://user-images.githubusercontent.com/96308/138757208-e625956a-409a-488c-85d8-372f7580274b.png)|

# Follow-up

I'd like to add the second screen from https://github.com/WordPress/gutenberg/issues/31456#issuecomment-888218396 (where users can discard changes they've made) in a follow-up PR.

ockham committed:
EntityTypeList: Add entity descriptions

ockham committed:
Sort entity groups

ockham committed:
Change panel header copy

paaljoachim commented:
Testing method: https://make.wordpress.org/design/2021/03/03/testing-a-gutenberg-pull-request-pr/

This is what I see. (I happen to test out a General template called Test.)

![Screenshot 2021-10-25 at 21 49 14](https://user-images.githubusercontent.com/5323259/138760662-d5b04801-4497-41f1-8b82-6d125e1e8d54.jpg)

Here are some thoughts that show up as I go through these sections. 

Are you ready to save? (Kinda strange seeing that here as I have already clicked the Save button once. Next step is to go through the various areas listed. Make needed adjustments unclick (or some other way) to discard what I do not want to save and then click Save again to save the areas I want saved.)

"The following changes have been made to your site, templates, and content." 
(Is clear and to the point sentence.)

Site - "These changes will affect your whole site." (To the point and helpful.)

Template - "These changes will affect pages and posts that use these templates." (To the point and helpful.)

Template Part (should we have a sentence here?) Something like: These changes will affect the following template parts. As in a real life situation header might be called header front page, header contact page, footer front page etc. 

Page - Home 
I need some clarity how a template called Test

![Screenshot 2021-10-25 at 22 03 43](https://user-images.githubusercontent.com/5323259/138762733-83887c1b-2b30-43c1-89f6-679263f31237.jpg)

affects the Page - Home. Similar to how further up the Template panel has this sentence: "These changes will affect pages and posts that use these templates."

Thanks.

ockham commented:
Thanks for reviewing, @paaljoachim! I've mostly tried to stick to the wording specified by the designs in https://github.com/WordPress/gutenberg/issues/31456#issuecomment-888218396 (with one minor change, I put "The following changes have been made to your site, templates, and content." rather than "The following changes have been made to your site, post, and templates." to reflect that the changes can affect any kind of content -- rather than just a post -- and to reflect the new entity order).

That said, I don't mind changing the copy; but I would like to check back with designers, as per my understanding, we need to strike a balance between providing enough explanatory texts to help the user understand what they're doing, but not too much as to not overwhelm them 🙂

ockham commented:
Investigating failing e2e tests 👀

ockham committed:
Remove undefined entries from partitioned savables

ockham commented:
> Investigating failing e2e tests 👀

Pushed a fix: 034eace

javierarce commented:
This works great, @ockham! 

I have just some minor suggestions for improvements:

- I think we could get rid of the icons next to the Site and Page labels. Maybe we could rethink the use of icons in a second iteration of the sidebar.
- Could we pluralize the messages depending on the number of changes? For example, if there's just one change, could we say: "This change will affect your whole site" instead of "These changes will affect your whole site"?
- I would capitalize "Header" and "Footer", like we do now in the template tab.
- I would also combine template and template parts into the "Templates" section. The concept of "template part" is vague, and this would simplify things. It would also solve the problem with the missing description that @paaljoachim mentioned. I would place the templates on top of the list and the parts on the bottom.

And here's something I just noticed when I saved a post after changing its title: we are currently showing the new title. Could we say "Title" as we do with the site?

<img width="276" alt="Screenshot 2021-10-27 at 17 05 52" src="https://user-images.githubusercontent.com/4933/139095258-e369bce0-6cfe-401f-a2c4-cbfa985e4fc0.png">


> Thanks for reviewing, @paaljoachim! I've mostly tried to stick to the wording specified by the designs in #31456 (comment) (with one minor change, I put "The following changes have been made to your site, templates, and content." rather than "The following changes have been made to your site, post, and templates." to reflect that the changes can affect any kind of content -- rather than just a post -- and to reflect the new entity order).

+1, much better than what I suggested.

ockham committed:
Use singular and plural depending on number of savables per entity

ockham committed:
Tweak singular

ockham commented:
Hey @javierarce, thanks a lot for reviewing!

I'll address each item in a separate comment, starting with one of the easiest ones:

> Could we pluralize the messages depending on the number of changes? For example, if there's just one change, could we say: "This change will affect your whole site" instead of "These changes will affect your whole site"?

Done in ef6f357c5b and d6b8679563:

![image](https://user-images.githubusercontent.com/96308/139103867-d4e70547-132d-469f-85a2-c56fb6d780fd.png)

ockham committed:
Remove icons from entities

ockham commented:
> I think we could get rid of the icons next to the Site and Page labels. Maybe we could rethink the use of icons in a second iteration of the sidebar.

88adb53

ockham commented:
> And here's something I just noticed when I saved a post after changing its title: we are currently showing the new title. Could we say "Title" as we do with the site?

That's not really specific to _title_ changes, though: It will refer to the modified post by title irrespective of _what_ part of the post was changed. (Try e.g. changing the post date through the corresponding block.) So I'd rather not change that 🤔

javierarce commented:
> That's not really specific to title changes, though: It will refer to the modified post by title irrespective of what part of the post was changed. (Try e.g. changing the post date through the corresponding block.) So I'd rather not change that 🤔

Oh, I didn't notice that, sorry. Ok, let's leave it then.

ockham committed:
Use template part area label rather than template part title

paaljoachim commented:
@javierarce 

Javier I believe you mentioned merging templates and template parts into one panel + capitalizing Header and Footer. Like so:

![Multi-save-suggestion](https://user-images.githubusercontent.com/5323259/139273112-48bf69f2-63c1-4f1b-a6c8-07757bdf21f1.jpg)


**NB! I added some text under the Pages panel.**
Because the text: "This change will affect pages and posts that use this template." is located inside the Template panel creating association to the specific panel. If the panel is closed the text will not be seen in the Page panel further below. I would suggest to add a description that builds upon the text seen under template. 
"The following pages use this template."

ockham commented:
> I would capitalize "Header" and "Footer", like we do now in the template tab.

This one was a bit trickier than it seems. The template part names are specified by the template part itself, and could theoretically be pretty much any string -- and in some cases maybe shouldn't be capitalized (think i18n).

What we seem to do in cases like this is find the template part _area_ a template part is meant for. For some of those (such as header and footer), we've defined specific icons and labels (which are properly capitalized, and -- hopefully! -- translated).

I've implemented this in 4ef8ac1:

![image](https://user-images.githubusercontent.com/96308/139273261-99092140-3315-4f0b-9dfa-9b2283bb91ec.png)

ockham commented:
> **NB! I added some text under the Pages panel.**
Because the text: "This change will affect pages and posts that use this template." is located inside the Template panel creating association to the specific panel. If the panel is closed the text will not be seen in the Page panel further below.

Hmm, this is fine, though: The text is really only meant to refer to templates. The Page panel however is about the page _content_ (rather than the template) that the user has modified.

> I would suggest to add a description that builds upon the text seen under template.
"The following pages use this template."

We can do that, but it's a bit misleading, and it won't scale too well:

It's technically correct since AFAICT, we'll only see this panel if we're editing the template for a specific page (rather than a generic template, such as _Single Post_ or _Index_). However, as stated above, the other necessary condition for seeing this panel is that we've modified the page _content_. So it would be more correct to say something like "The content of the following page has been modified."

However, if we allow editing content other than individual pages in the site editor -- e.g. multiple posts through a template with a query loop -- that description would no longer be accurate. How about something like "The following content has been modified." instead?

paaljoachim commented:
This sounds good to me. What do you think @javierarce ?

> How about something like "The following content has been modified." instead?

I assume we should have @kellychoffman and @jameskoster check out the PR as well.

ockham commented:
> I would also combine template and template parts into the "Templates" section. The concept of "template part" is vague, and this would simplify things. It would also solve the problem with the missing description that @paaljoachim mentioned. I would place the templates on top of the list and the parts on the bottom.

This item is the trickiest one. I'm not totally sure I agree that having templates and template under the same heading simplifies thing 😬 On the development side, it would actually be a rather big change of the relevant code. I'd just like to double-check this is really what we want before I make that change 😅 

From a UX-perspective, it was my impression that we have a few invariants that differentiate both concepts:

- A template affects an entire "screen" of what a user sees -- be it on the frontend or in the editor. As a corollary, the site editor always only shows exactly one template. By conclusion, the multi-entity Save UI will never show more than one template as affected by user changes.
- A template part defines part of a "screen" that can be used across templates, typically to provide some visual consistency, and reusability. E.g. a header or footer template part that's used across all templates of a theme. A template can contain any number of template parts; any of them can be modified by the user in the site editor.

javierarce commented:
> I'm not totally sure I agree that having templates and template under the same heading simplifies thing 😬 On the development side, it would actually be a rather big change of the relevant code. I'd just like to double-check this is really what we want before I make that change 😅

After thinking this through again and reading your explanation, I feel that it could be confusing to list the name of a template (say 404) and then "header" and "footer" immediately below. 

For me, templates and template parts are conceptually very similar (they are structures), but the difference in size and side-effects at that part of the flow could be disorienting for a user who has made many changes. 

The solution that @paaljoachim designed [in this comment](https://github.com/WordPress/gutenberg/pull/35933#issuecomment-953879288) sits in a good middle ground that could simplify the sidebar and at the same time maintain that differentiation. The only thing I would change is the "Template part" label: I would use "Areas" as we do in the header dropdown.

ntsekouras commented on packages/editor/src/components/entities-saved-states/entity-type-list.js:
```diff
@@ -34,12 +44,13 @@ export default function EntityTypeList( {
 	return (
-		<PanelBody title={ entity.label } initialOpen={ true } icon={ icon }>
+		<PanelBody title={ entity.label } initialOpen={ true }>
+			{ description ? <PanelRow>{ description }</PanelRow> : null }
```
```suggestion
			{ description && <PanelRow>{ description }</PanelRow> }
```

ntsekouras commented on packages/editor/src/components/entities-saved-states/entity-record-item.js:
```diff
@@ -47,9 +50,25 @@ export default function EntityRecordItem( {
+				).title;
+			}
+
+			if ( name === 'wp_template_part' ) {
```
When we have more than one template parts from the same area, this can become confusing, no?

Mamaduka replied:
I agree it can be confusing. Users might have more than one template from the "General" area. You can use the "Armando" theme to test this.

Maybe we should use the actual template part title and fall back to a name. Now that we fully support defining titles in theme.json - #35626.

ockham replied:
Ah, thanks, I was unaware of #35626! Happy to use the template part title instead 😄 

> Maybe we should use the actual template part title and fall back to a name.

Not sure I'm totally getting you right, shouldn't the title field (as supplied by the endpoint) be doing that already, per https://github.com/WordPress/gutenberg/pull/35626/files#diff-6ca30c26729f0fb3398ad5c6b4e82d1a83bfe5bd920e4cb6deca5646d9e578bdR226? 🤔

Mamaduka replied:
Correct, we're doing fallback on the server-side 😄

ockham replied:
I've reverted 4ef8ac1 which seems to be enough to get the desired behavior. (I've verified by adding a `title` to my local tt1-blocks' `theme.json`, and deleting the existing template part CPTs.)

cc/ @javierarce

ntsekouras commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -68,9 +68,21 @@ export default function EntitiesSavedStates( { close } ) {
+		templateSavables,
+		templatePartSavables,
+		...Object.values( contentSavables ),
+	].filter( Array.isArray );
```
Can we just assign a default value (empty array) at the destructuring instead of calling `filter`?

ockham replied:
I was considering that initially, but I think we'd then need some extra guarding logic [around here](https://github.com/WordPress/gutenberg/pull/35933/files#diff-c5815980e7aa19efd287aaa72512569ac7de514d74679af2f0b099f8a9369bc5R183) (or maybe inside of `EntityTypeList`), since we'd otherwise still create an `<EntityTypeList />` component for each item of `sortedPartitionedSavables`, even if it that item itself is only an empty array.

ntsekouras reviewed (commented):
Thanks for working on this @ockham!

This is looking good code wise and my sole concern is the `template parts` area comment.

ockham committed:
Revert "Use template part area label rather than template part title"

This reverts commit 4ef8ac11d63df3716a349ecf85528c48ebde1cec.

ockham committed:
Replace ternary by &&

Co-authored-by: Nik Tsekouras <ntsekouras@outlook.com>

ockham committed:
Add description for content entities

ntsekouras commented on packages/editor/src/components/entities-saved-states/entity-type-list.js:
```diff
@@ -6,20 +6,35 @@ import { some } from 'lodash';
+		),
+	};
+
+	return (
```
I guess this can become a bit overwhelming with many changes 😄
<img width="881" alt="Screenshot 2021-10-29 at 4 15 48 PM" src="https://user-images.githubusercontent.com/16275880/139442341-12b5e887-a1b1-48ce-a049-06f952ff4f23.png">

ockham replied:
That was partly by mistake -- there shouldn't be any description for Template Part. Fix in 717d7041ba.

I don't mind removing the "The following content has been modified." description altogether, if it's still too overwhelming.

ntsekouras replied:
I personally like it for the page edits. Not sure for reusable blocks though, but not a blocker.

ockham replied:
Ah, I'd somehow missed the reusable blocks. Good point, I'll do some final tweaking...

ockham committed:
No description for Template Part

ockham committed:
Singular/plural for Template Part(s) header

ockham commented:
> The solution that @paaljoachim designed in this comment sits in a good middle ground that could simplify the sidebar and at the same time maintain that differentiation.

Thanks @javierarce! I believe that the latest couple of commits should've tweaked the UI accordingly (see image at the bottom of this comment).

> The only thing I would change is the "Template part" label: I would use "Areas" as we do in the header dropdown.

In line with the outcome of my [discussion](https://github.com/WordPress/gutenberg/pull/35933#discussion_r739060839) with @ntsekouras and @Mamaduka, I've kept this as "Template part", since we really show the template parts that are included with a theme, and also use their titles (rather than the areas they're associated with) -- which could be something like "Default header", "Alternative header", "Big footer", "Small footer" etc, since a given theme can theoretically provide multiple template parts for a given area.

What I've done however is pluralize correctly if there are multiple template part**s** affected:

![image](https://user-images.githubusercontent.com/96308/139443228-ba64d370-fd72-4d38-8d98-75a1c15a7fa3.png)

ntsekouras reviewed (approved):
LGTM, thanks @ockham !

Mamaduka commented:
Thanks for working on this, @ockham.

paaljoachim reviewed (commented):
The design looks good to me! Thank you for working on this PR @ockham Bernie!

Here is a before and after.
![Multisave](https://user-images.githubusercontent.com/5323259/139454289-1ad03d0b-ee8c-46dc-b685-f9679a328e10.jpg)

javierarce commented:
Thanks for working on this, @ockham. The changes look good!

ockham commented:
Thanks all for your feedback and reviews! ❤️

ockham committed:
Tweak entity descriptions to exclude reusable blocks

youknowriad commented on packages/editor/src/components/entities-saved-states/entity-type-list.js:
```diff
@@ -6,20 +6,35 @@ import { some } from 'lodash';
+			);
+		case 'page':
+		case 'post':
+			return __( 'The following content has been modified.' );
```
Should we move these descriptions to the entities config instead?

ockham replied:
You mean https://github.com/WordPress/gutenberg/blob/ba5f6d061934625027beab9d2bf600ae42a63be9/packages/core-data/src/entities.js? Yeah, that might make sense 👍

youknowriad commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -68,9 +68,21 @@ export default function EntitiesSavedStates( { close } ) {
+	const {
+		site: siteSavables,
+		wp_template: templateSavables,
+		wp_template_part: templatePartSavables,
```
Do you think that order can be defined as a "priority" somehow in the entities config? Or maybe this is sufficient for now?

ockham replied:
I think it'd make the code quite a bit more complex, and I don't think it's worth it right now, since this is the only place so far where we're needing that priority order.
